### PR TITLE
imp: Implement custom JSON and Borsh deserialization for `ChainId`

### LINKED ISSUE #996: Code accepts inconsistent serialised ChainId objects
## Bug Summary

It’s possible to craft JSON or Borsh serialised object which deserialises into an inconsistent ChainId object which has invalid revision_number field.

## Demonstration

```rust
use borsh::BorshDeserialize;
use ibc_core_host_types::identifiers::ChainId;

fn evil_borsh() -> ChainId {
    ChainId::try_from_slice(b"\x06\0\0\0foo-42\x45\0\0\0\0\0\0\0").unwrap()
}

fn evil_json() -> ChainId {
    serde_json::from_str(r#"{"id": "foo-42", "revision_number": 69}"#).unwrap()
}

fn show(id: ChainId) {
    println!("id: {}; revision: {}", id.as_str(), id.revision_number());
        // → id: foo-42; revision: 69
    let other = ChainId::new(id.as_str()).unwrap();
    println!("id: {}; revision: {}", other.as_str(), other.revision_number());
        // → id: foo-42; revision: 42
    // assert_eq!(id, other); // → panics
}

fn main() {
    show(evil_borsh());
    show(evil_json());
}
```

## Solution

Either serialisation format should change or deserialisation should verify the object is valid.

IMO ideally ChainId would be serialised as a String without the revision_number and the revision would be filled in when deserialising.  Alas, this would be an ABI breaking change.

ABI-stable alternative is for deserialise implementations to verify that the parsed object is correct.

(Side note: IMO revision_number shouldn’t even be a field in ChainId and instead it should be parsed when requested but that’s a separate matter).

## Version

```toml
ibc-core-host-types = { version = "0.48.1", features = ["serde", "borsh"] }
```

### PR DESCRIPTION
**Pull request by seanchen1991**
Closes: #996 

## Description

______

### PR author checklist:
- [ ] Added changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [ ] Added tests.
- [ ] Linked to GitHub issue.
- [ ] Updated code comments and documentation (e.g., `docs/`).
- [ ] Tagged *one* reviewer who will be the one responsible for shepherding this PR.

### Reviewer checklist:

- [ ] Reviewed `Files changed` in the GitHub PR explorer.
- [ ] Manually tested (in case integration/unit/mock tests are absent).

### CODE REVIEW

File: ibc-core/ics24-host/types/src/identifiers/chain_id.rs [hunk 1]
```diff
@@ -116,6 +115,173 @@ impl ChainId {
...
+                        write!(formatter, "expected one of: {:?}", &FIELDS)
+                    }
+
+                    fn visit_str<E>(self, value: &str) -> Result<Field, E>
+                    where
+                        E: Error,
+                    {
+                        match value {
+                            "id" => Ok(Field::Id),
+                            "revisionnumber" | "revision_number" => Ok(Field::RevisionNumber),
```
Comment (Farhad-Shabani) [inline ibc-core/ics24-host/types/src/identifiers/chain_id.rs]: Is this `revisionnumber` necessary?

File: ibc-core/ics24-host/types/src/identifiers/chain_id.rs [hunk 2]
```diff
@@ -116,6 +115,173 @@ impl ChainId {
...
+                E: Error,
+            {
+                match ChainId::from_str(v) {
+                    Ok(chain_id) => Ok(chain_id),
+                    Err(e) => Err(Error::custom(format_args!(
+                        "invalid chain ID `{}`: `{}`",
+                        v, e
+                    ))),
+                }
+            }
```
Comment (Farhad-Shabani) [inline ibc-core/ics24-host/types/src/identifiers/chain_id.rs]: I think this `visit_str` can be dropped. Only with the `visit_map` desrialization works as intended.

File: ibc-core/ics24-host/types/src/identifiers/chain_id.rs [hunk 3]
```diff
@@ -116,6 +115,173 @@ impl ChainId {
...
+                                }
+                                _ => {
+                                    validate_identifier_length(next_value, 1, 64).map_err(
+                                        |_| Error::custom("invalid chain ID identifier length"),
+                                    )?;
+
+                                    id = Some(next_value);
+                                    revision_number = Some(0);
+                                }
+                            }
```
Comment (Farhad-Shabani) [inline ibc-core/ics24-host/types/src/identifiers/chain_id.rs]: `validate_identifier_chars(next_value)` should be performed before this match statement. This part can be simplified by using `ChainId::from_str` instead, as follows:

```suggestion
                            let chain_id =
                                match ChainId::from_str(next_value).map_err(Error::custom) {
                                    Ok(chain_id) => chain_id,
                                    Err(e) => return Err(e),
                                };
                            id = Some(chain_id.id);
                            revision_number = Some(chain_id.revision_number)
```

Good to have a test case for a `ChainId` with invalid chars e.g. `#[case(r#"{"id":"/foo-42","revision_number":"0"}"#)]` must fail.

File: ibc-core/ics24-host/types/src/identifiers/chain_id.rs [hunk 4]
```diff
@@ -116,6 +115,173 @@ impl ChainId {
...
+        revision_number: u64,
+    }
+
+    impl BorshDeserialize for ChainId {
+        fn deserialize_reader<R: Read>(reader: &mut R) -> io::Result<Self> {
+            let inner = InnerChainId::deserialize_reader(reader)?;
+
+            let Ok((_, rn)) = parse_chain_id_string(&inner.id) else {
+                return Err(Error::new(ErrorKind::Other, "failed to parse chain ID"));
+            };
```
Comment (Farhad-Shabani) [inline ibc-core/ics24-host/types/src/identifiers/chain_id.rs]: Despite not being parsed, a chain id string is not necessarily invalid. The `#[case(r#"{"id":"foo","revision_number":"0"}"#)]` should be successfully deserialized. 
Could you add tests for valid borsh desrialization cases as well?

File: ibc-core/ics24-host/types/src/identifiers/chain_id.rs [hunk 5]
```diff
@@ -250,10 +416,35 @@ mod tests {
...
     }
 
+    #[cfg(feature = "serde")]
+    #[rstest]
+    #[case(r#"{"id":"foo-42","revision_number":"42"}"#)]
+    #[case(r#"{"id":"foo-42","revision_number":"0"}"#)]
+    fn test_valid_chain_id_json_deserialization(#[case] chain_id_json: &str) {
+        let chain_id = serde_json::from_str::<ChainId>(chain_id_json);
+        std::println!("{:?}", chain_id);
+        assert!(chain_id.is_ok());
```
Comment (Farhad-Shabani) [inline ibc-core/ics24-host/types/src/identifiers/chain_id.rs]:
```suggestion
```

File: ibc-core/ics24-host/types/expanded.rs [hunk 6]
```diff
@@ -0,0 +1,3890 @@
+#![feature(prelude_import)]
```
Comment (Farhad-Shabani) [inline ibc-core/ics24-host/types/expanded.rs]: This `expanded.rs` file looks irrelevant and can be removed.

File: .changelog/unreleased/improvements/996-custom-chain-id-deserialization.md [hunk 7]
```diff
@@ -0,0 +1 @@
+- `[ibc-core]` Implement custom JSON and Borsh deserialization for `ChainId` ([#996](https://github.com/cosmos/ibc-rs/pull/1013))
```
Comment (Farhad-Shabani) [inline .changelog/unreleased/improvements/996-custom-chain-id-deserialization.md]:
```suggestion
- `[ibc-core-host-types]` Implement custom JSON and Borsh deserialization for `ChainId` ([#996](https://github.com/cosmos/ibc-rs/pull/1013))
```

**Review by Farhad-Shabani — CHANGES_REQUESTED** [review]

File: ibc-core/ics24-host/types/src/identifiers/chain_id.rs [hunk 8]
```diff
@@ -116,6 +115,173 @@ impl ChainId {
     }
 }
 
+#[cfg(feature = "serde")]
+impl<'de> Deserialize<'de> for ChainId {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        const FIELDS: &[&str] = &["id", "revision_number"];
```
Comment (mina86) [inline ibc-core/ics24-host/types/src/identifiers/chain_id.rs:124]: Why not something like: 

```rust
#[derive(serde::Deserialize)]
struct IdAndRevision {
    id: String,
    revision_number: u64,
}

let IdAndRevision { id, revision_number } =
    IdAndRevision::deserialize(D)?;
if revision_number == parse_revision_number(chain_id.as_str())? {
    Ok(Self { id, revision_number })
} else {
    Err(...)
}
```

And add `fn parse_revision_number(id: &str) -> Result<u64, IdentifierError>;` function

File: ibc-core/ics24-host/types/src/identifiers/chain_id.rs [hunk 9]
```diff
@@ -116,6 +115,173 @@ impl ChainId {
...
+
+    #[derive(BorshDeserialize)]
+    pub struct InnerChainId {
+        id: String,
+        revision_number: u64,
+    }
+
+    impl BorshDeserialize for ChainId {
+        fn deserialize_reader<R: Read>(reader: &mut R) -> io::Result<Self> {
+            let inner = InnerChainId::deserialize_reader(reader)?;
```
Comment (mina86) [inline ibc-core/ics24-host/types/src/identifiers/chain_id.rs]: I’d suggest

```rust
let (id, revision_number) = 
    <(String, u64)>::deserialize_reader(reader)?;
```

and then there’s no need for the `InnerChainId` type.

File: ibc-core/ics24-host/types/src/identifiers/chain_id.rs (hunk 4 above)
Comment (mina86) [reply ibc-core/ics24-host/types/src/identifiers/chain_id.rs]: What about `{"id":"foo-bar-42","revision_number":"0"}`? The condition below accepts it but my intuition is that it should be rejected as well. IMO deserialised object should be the same as one constructed via `ChainId::from_str(&deserialised.id)`.

File: ibc-core/ics24-host/types/src/identifiers/chain_id.rs (hunk 1 above)
Comment (seanchen1991) [reply ibc-core/ics24-host/types/src/identifiers/chain_id.rs]: Yes, for some reason, both are needed. Not really sure why, to be honest, but removing one or the other will break some tests.

File: ibc-core/ics24-host/types/src/identifiers/chain_id.rs (hunk 3 above)
Comment (seanchen1991) [reply ibc-core/ics24-host/types/src/identifiers/chain_id.rs]: [50aa991](https://github.com/cosmos/ibc-rs/pull/1013/commits/50aa991cc362da69cd749300fabcdd01c9df6c20)

File: ibc-core/ics24-host/types/src/identifiers/chain_id.rs (hunk 2 above)
Comment (seanchen1991) [reply ibc-core/ics24-host/types/src/identifiers/chain_id.rs]: [50aa991](https://github.com/cosmos/ibc-rs/pull/1013/commits/50aa991cc362da69cd749300fabcdd01c9df6c20)

File: .changelog/unreleased/improvements/996-custom-chain-id-deserialization.md (hunk 7 above)
Comment (seanchen1991) [reply .changelog/unreleased/improvements/996-custom-chain-id-deserialization.md]: [e92fa78](https://github.com/cosmos/ibc-rs/pull/1013/commits/e92fa7820982af117d90d25c698282868cf0d507)

File: ibc-core/ics24-host/types/expanded.rs (hunk 6 above)
Comment (seanchen1991) [reply ibc-core/ics24-host/types/expanded.rs]: [e92fa78](https://github.com/cosmos/ibc-rs/pull/1013/commits/e92fa7820982af117d90d25c698282868cf0d507)

File: ibc-core/ics24-host/types/src/identifiers/chain_id.rs (hunk 5 above)
Comment (seanchen1991) [reply ibc-core/ics24-host/types/src/identifiers/chain_id.rs]: [e92fa78](https://github.com/cosmos/ibc-rs/pull/1013/commits/e92fa7820982af117d90d25c698282868cf0d507)

File: ibc-core/ics24-host/types/src/identifiers/chain_id.rs (hunk 8 above)
Comment (seanchen1991) [reply ibc-core/ics24-host/types/src/identifiers/chain_id.rs:124]: It's not really clear to me which part of the `serde::Deserialize` impl you're thinking your suggested code should replace.

File: ibc-core/ics24-host/types/src/identifiers/chain_id.rs (hunk 9 above)
Comment (seanchen1991) [reply ibc-core/ics24-host/types/src/identifiers/chain_id.rs]: [b1063e6](https://github.com/cosmos/ibc-rs/pull/1013/commits/b1063e6d45ef6ff838f85978535df0e3557a8e14)

File: ibc-core/ics24-host/types/src/identifiers/chain_id.rs (hunk 4 above)
Comment (seanchen1991) [reply ibc-core/ics24-host/types/src/identifiers/chain_id.rs]: It's possible to have `ChainId`s that don't strictly follow the format, but are still considered valid. `{"id":"foo-bar-42","revision_number":"0"}` is an example of a `ChainId` that is valid but that doesn't strictly follow the format.

File: ibc-core/ics24-host/types/src/identifiers/chain_id.rs (hunk 8 above)
Comment (mina86) [reply ibc-core/ics24-host/types/src/identifiers/chain_id.rs:124]: This would be body of the `deserialize` method:

```rust
#[cfg(feature = "serde")]
impl<'de> Deserialize<'de> for ChainId {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        #[derive(Deserialize)]
        struct IdAndRevision {
            id: String,
            #[serde(alias = "revisionnumber")]
            revision_number: u64,
        }

        let IdAndRevision { id, revision_number } =
            IdAndRevision::deserialize(D)?;
        if revision_number == parse_revision_number(chain_id.as_str())? {
            Ok(Self { id, revision_number })
        } else {
            Err(...)
        }
    }
}
```

Comment (mina86) [reply ibc-core/ics24-host/types/src/identifiers/chain_id.rs:124]: What I’ve posted would be body of the `deserialize` method, i.e.:

```rust
#[cfg(feature = "serde")]
impl<'de> Deserialize<'de> for ChainId {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        #[derive(Deserialize)]
        struct IdAndRevision {
            id: String,
            revision_number: u64,
        }

        let IdAndRevision { id, revision_number } =
            IdAndRevision::deserialize(deserializer)?;
        if revision_number == 0 {
            // If `revision_number` is zero we allow whatever valid identifier.
            // In particular, `{id: "foo-bar-42", revision_number: 0}` is valid.
            validate_chain_id(id.as_str())?;
        } else if revision_number != parse_revision_number(id.as_str())? {
            return Err(...);
        }
        Ok(Self { id, revision_number })
    }
}
```

Comment (Farhad-Shabani) [reply ibc-core/ics24-host/types/src/identifiers/chain_id.rs:124]: Still `IdAndRevision::deserialize(deserializer)?` is not working properly for all cases. Anyhow let’s leave further optimization for later PRs.

**Review by Farhad-Shabani — APPROVED** [review]
Thanks @seanchen1991 for sorting this out, and a nod to @mina86 for chiming in!

File: ibc-core/ics24-host/types/src/identifiers/chain_id.rs (hunk 8 above)
Comment (mina86) [reply ibc-core/ics24-host/types/src/identifiers/chain_id.rs:124]: What’s the issue? IdAndRevision with a derive should have the exact same deserialisation logic as ChainId.
